source/common/common/hash.h

```diff
@@ -26,6 +27,42 @@ class HashUtil {
     return XXH64(input.data(), input.size(), seed);
   }
 
+  /**
+   * Return 64-bit hash from deterministically serializing a value in
+   * an endian-independent way and hashing it with the xxHash algorithm.
+   *
+   * Floats and doubles are just done as endian-adjusted raw bytes the same
+   * as integers; this is on par with what protobuf serialization does.
+   * src/google/protobuf/generated_message_util.cc literally just treats
+   * a float the same as int32 for serialization purposes.
```

Comment:
So all the bits are *something*, but maybe "NaN with a tag" should not be hashed differently than "NaN without a tag", this is not particularly clear to me. I have changed it to mantissa+exponent based hashing for floating point, which sacrifices a little performance to gain a little predictability (and added tests specifically for weird NaN, which would fail with the old version, but again, maybe the test is wrong to require these to hash the same!)